src/lib/operators/join_hash.hpp

```diff
@@ -52,6 +52,13 @@ class JoinHash : public AbstractJoinOperator {
     size_t radix_bits{0};
     // Initially, the left input is the build side and the right side is the probe side.
     bool left_input_is_build_side{true};
+
+    // Due to the used Bloom filters, the number of actually joined tuples can significantly differ from the sizes of
+    // the input tables. For the build side, the bloom filter reduces the number of values in the hash table (i.e., the
+    // size of the hash table) and potentially the number of rows (in case of non-semi/anti* joins).
+    size_t probe_side_materialized_values{0};
+    size_t build_side_hash_table_size{0};
+    size_t build_side_position_count{0};
```

Comment:
Are you making the assumption that the build side is always materialized first and the probe side is the first that is filtered by a bloom filter? If yes, this does not seem to match the JoinHash implementation:

https://github.com/hyrise/hyrise/blob/adaa9d5305eb952d4125cd66b20f5e58ef62a9eb/src/lib/operators/join_hash.cpp#L359-L372

Reply:
I just realized that the remainder of the hash join code largely works under the assumption that the if statement from line 359 is true.

Unless I am mistaken, if we materialize the probe side first, it does not make any sense to use a bloom filter when building the hash table. On the other side, we *could* use a bloom filter in the probe methods if we have not yet filtered the probe side when materializing it. This seems to catch some additional cases, but at least for TPC-H, I did not see a significant gain.

Reply:
> Are you making the assumption that the build side is always materialized first and the probe side is the first that is filtered by a bloom filter?

Nope. That's also what's being tested (next comment).

Concerning your other comment: I thought about it. But since we invest a lot of work for L2 local hash tables, I thought (might be totally wrong) that the saved hash lookup does not make up the costs of the Bloom lookup. To be honest, I don't have the time to look into it. I could add a comment about it though.

Reply:
Ah, are you talking about this?

> As we first materialize the smaller side (i.e., the probe side in this case) and create the initial bloom filter with that, there will be no reduction due to bloom filtering on the probe side.

But the build side would be bloom-filtered, right? Why don't we have a `build_side_materialized_value_count`?

Reply:
> Concerning your other comment: I thought about it. But since we invest a lot of work for L2 local hash tables, I thought (might be totally wrong) that the saved hash lookup does not make up the costs of the Bloom lookup. To be honest, I don't have the time to look into it. I could add a comment about it though.

This was my impression, too. In some cases, it seemed to hurt the performance. A comment would certainly help.

Similarly, I wonder if we go down the `else` part of the quoted statement (i.e., materialize the probe side first and bloom-filter the build side), it makes any sense to pass in the bloom filter here:

https://github.com/hyrise/hyrise/blob/adaa9d5305eb952d4125cd66b20f5e58ef62a9eb/src/lib/operators/join_hash.cpp#L435-L436

Again, no need to change anything, maybe just add a comment if you agree.

Reply:
> Why don't we have a build_side_materialized_value_count?

The current "naming scheme" is meant to clearly differentiate between the two kind of "values" we are interested in:
  1. Distinct values in the hash map (i.e., size), when no positions are stored (semi/anti case)
  1. Distinct positions in the values of the hash table's entries, that's basically the number of remaining table tuples after bloom-filtering (inner case).

`build_side_materialized_value_count` does not clearly state what is means, I think.

Will add comment about the two issues mentioned.

Reply:
Added
```
// Here, we first materialize the probe side and use the resulting Bloom filter in the build side
// materialization. Consequently, the passed Bloom filter in build() will have no effect.
```